File: Data Structures & Algorithms/house-robber/submission-0.java
class Solution {
    int dp[];
    private int find(int[] nums,int i,int n) {
        if(i >= n) {
            return 0;
        }
        if(dp[i] > 0) return dp[i];
        int result = Math.max(nums[i] + find(nums, i + 2, n), find(nums, i + 1, n));
        dp[i] = result;
        return result;
    }
    public int rob(int[] nums) {
        int n = nums.length;
        dp = new int[n];
        return find(nums, 0, n);
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    constructor() {
    }
    find(nums, curr, prev) {
        if(curr < 0) {
            return 0;
        }
        const nottake = this.find(nums, curr-1, prev);
        let take = 0;
        if(prev === -1 || nums[curr] < nums[prev]) {
            take = 1 + this.find(nums, curr-1, curr);
        }
        return Math.max(nottake, take);
    }
    lengthOfLIS(nums) {
        const n = nums.length;
        return this.find(nums, n-1, -1);
    }
}
